kratos/python/add_properties_to_python.cpp

```diff
@@ -254,70 +283,22 @@ void  AddPropertiesToPython(pybind11::module& m)
-            { return rProperties.GetAccessor(rVariable); },
-            py::return_value_policy::reference_internal)
-        .def(
-            "GetAccessor", [](Properties &rProperties, const Variable<array_1d<double, 3>> &rVariable) -> auto&
-            { return rProperties.GetAccessor(rVariable); },
-            py::return_value_policy::reference_internal)
-        .def(
-            "GetAccessor", [](Properties &rProperties, const Variable<array_1d<double, 6>> &rVariable) -> auto&
-            { return rProperties.GetAccessor(rVariable); },
-            py::return_value_policy::reference_internal)
-        .def(
-            "GetAccessor", [](Properties &rProperties, const Variable<array_1d<double, 9>> &rVariable) -> auto&
-            { return rProperties.GetAccessor(rVariable); },
-            py::return_value_policy::reference_internal)
-        .def(
-            "GetAccessor", [](Properties &rProperties, const Variable<array_1d<double, 4>> &rVariable) -> auto&
-            { return rProperties.GetAccessor(rVariable); },
-            py::return_value_policy::reference_internal)
-        .def(
-            "GetAccessor", [](Properties &rProperties, const Variable<std::string> &rVariable) -> auto&
-            { return rProperties.GetAccessor(rVariable); },
-            py::return_value_policy::reference_internal)
-
-        .def("HasAccessor", [](Properties &rProperties, const Variable<bool> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<int> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<double> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<Vector> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<Matrix> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<array_1d<double, 3>> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<array_1d<double, 6>> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<array_1d<double, 9>> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<array_1d<double, 4>> &rVariable)
-             { return rProperties.HasAccessor(rVariable); })
-        .def("HasAccessor", [](Properties &rProperties, const Variable<std::string> &rVariable)
-             { return rProperties.HasAccessor(rVariable); });
+        ;
+
+        AddInterfaceToAccessor<
```

Comment:
won't it take forever to compile a function with so many template arguments? maybe alternatively we could use a few macros here.

Reply:
No, this is not a normal template. This is a folding expression. In general compilation performance of folding expression is waaaay better than recursive templates (which would be another way to do this, and performance the reason we werent using it)

Reply:
but why cannot you write

 ~~~cpp
AddInterfaceToAccessorFold<double>(...)
AddInterfaceToAccessorFold<bool>(...)
AddInterfaceToAccessorFold<int>(...)
AddInterfaceToAccessorFold<unsifned int>(...)
...
 ~~~

which would be a much simpler code?

Reply:
Well, yes I could, but then I have to write one line with all the args per type I want to expose. Imo the other way is better (note that we could have a macro with a type bundle (ie KRATOS_PYTHON_SCALAR_VARS, KRATOS_PYTHON_VECTOR_VARS) and  then it will look as:

```c++
#define KRATOS_PYTHON_SCALAR_VARS bool, int, double, string
#define KRATOS_PYTHON_VECTOR_VARS matrix, vector, array_1d<double, 3>, array_1d<double, 4>, array_1d<double, 6>, .... 
```

```c++
AddInterfaceToAccessor<
    KRATOS_PYTHON_SCALAR_VARS,
    KRATOS_PYTHON_VECTOR_VARS 
>(properties_module);
```
 
Also if we leave it as I did, then when templated lambdas become usable we will be able to do something like:

```c++
AddInterfaceToAccessorFold<T1,T2...TN>(args, []<template TX>(arg) {
     return myclass<TX>(args)
})
```

So we don't need the auxiliar function, the folder is hidden and all the interface is written once and looks like any other lambda export from pybind.